[Docs][KubeRay] add a guide for deploying vLLM with RayService (#47038)
ray-project/ray: doc/source/cluster/kubernetes/examples/vllm-rayservice.md

(kuberay-vllm-rayservice-example)=

# Serve a Large Language Model with vLLM on Kubernetes

This guide demonstrates how to [Serve a Large Language Model with vLLM](https://docs.ray.io/en/latest/serve/tutorials/vllm-example.html) on Kubernetes using KubeRay. The example in this guide deploys the `meta-llama/Meta-Llama-3-8B-Instruct` model from Hugging Face on Google Kubernetes Engine (GKE).

## Prerequisites

> This example downloads model weights from Hugging Face. You will need to complete the following

Author: done

prerequisites to successfully complete this guide:
* A [Hugging Face account](https://huggingface.co/)
* A Hugging Face [access token](https://huggingface.co/docs/hub/security-tokens) with read access to gated repos.

> * Access to the Llama 3 8B model. This usually requires signing an agreement on Hugging Face to access this model. Go to the [Llama 3 model page](https://huggingface.co/meta-llama/Meta-Llama-3-8B) for more details.

Author: done

## Create a Kubernetes cluster on GKE

Create a GKE cluster with a GPU node pool:
```sh
gcloud container clusters create kuberay-gpu-cluster \
    --machine-type=g2-standard-24 \
    --location=us-east4-c \
    --num-nodes=2 \
    --accelerator=type=nvidia-l4,count=2,gpu-driver-version=latest
```

> This example uses L4 GPUs. Each model replica will use 2 L4 GPUs using vLLM's tensor parallelism.

Reviewer: ```suggestion
This example uses L4 GPUs. Each model replica uses 2 L4 GPUs using vLLM's tensor parallelism.
```

Author: done

## Install the KubeRay Operator

Follow [Deploy a KubeRay operator](kuberay-operator-deploy) to install the latest stable KubeRay operator from the Helm repository.
The KubeRay operator Pod must be on the CPU node if you set up the taint for the GPU node pool correctly.

## Create a Kubernetes Secret containing your Hugging Face access token

Create a Kubernetes Secret containing your Hugging Face access token:
```sh
export HF_TOKEN=<Hugging Face access token>
kubectl create secret generic hf-secret   --from-literal=hf_api_token=${HF_TOKEN}   --dry-run=client -o yaml | kubectl apply -f -
```

> This secret will be referenced as an environment variable in the RayCluster used in the next steps.

Reviewer: ```suggestion
This guide references this secret as an environment variable in the RayCluster in the next steps.
```

## Deploy a RayService

Create a RayService custom resource:
```
kubectl apply -f https://raw.githubusercontent.com/ray-project/kuberay/master/ray-operator/config/samples/vllm/ray-service.vllm.yaml
```

> The RayService is configured to deploy a Ray Serve application, running vLLM as the serving engine for the Llama 3 8B Instruct model. The code used in this example can be found [here](https://github.com/ray-project/kuberay/blob/master/ray-operator/config/samples/vllm/serve.py).

Reviewer: ```suggestion
This step configures RayService to deploy a Ray Serve app, running vLLM as the serving engine for the Llama 3 8B Instruct model. You can find the code for this example [on GitHub](https://github.com/ray-project/kuberay/blob/master/ray-operator/config/samples/vllm/serve.py).
```

Author: done

You can inspect the Serve Config for more details about the Serve deployment:
```yaml
  serveConfigV2: |
    applications:
    - name: llm
      route_prefix: /
      import_path:  ray-operator.config.samples.vllm.serve:model
      deployments:
      - name: VLLMDeployment
        num_replicas: 1
        ray_actor_options:
          num_cpus: 8
          # NOTE: num_gpus is set automatically based on TENSOR_PARALLELISM
      runtime_env:
        working_dir: "https://github.com/ray-project/kuberay/archive/master.zip"
        pip: ["vllm==0.5.4"]
        env_vars:
          MODEL_ID: "meta-llama/Meta-Llama-3-8B-Instruct"
          TENSOR_PARALLELISM: "2"
```

> Wait for the RayService resource to be ready. You can inspect it's status by running the following command:

Reviewer: ```suggestion
Wait for the RayService resource to be ready. You can inspect its status by running the following command:
```

Author: done

```
$ kubectl get rayservice llama-3-8b -o yaml
```

The output should contain the following:
```
status:
  activeServiceStatus:
    applicationStatuses:
      llm:
        healthLastUpdateTime: "2024-08-08T22:56:50Z"
        serveDeploymentStatuses:
          VLLMDeployment:
            healthLastUpdateTime: "2024-08-08T22:56:50Z"
            status: HEALTHY
        status: RUNNING
```

## Send a prompt

> Once you've confirmed the Ray Serve deployment is healthy, you can establish a port-forwarding session for the Serve application:

Reviewer: ```suggestion
Confirm the Ray Serve deployment is healthy, then you can establish a port-forwarding session for the Serve app:
```

Author: done

```sh
$ kubectl port-forward svc/llama-3-8b-serve-svc 8000
```

> Note that this Kubernetes Service will be created after the Serve applications are ready and running.

Reviewer: ```suggestion
Note that KubeRay creates this Kubernetes Service after the Serve apps are ready and running.
```

## Prerequisites

This example downloads model weights from Hugging Face. You will need to complete the following
prerequisites to successfully complete this guide:
* A [Hugging Face account](https://huggingface.co/)
* A Hugging Face [access token](https://huggingface.co/docs/hub/security-tokens) with read access to gated repos.
* Access to the Llama 3 8B model. This usually requires signing an agreement on Hugging Face to access this model. Go to the [Llama 3 model page](https://huggingface.co/meta-llama/Meta-Llama-3-8B) for more details.

## Create a Kubernetes cluster on GKE

Create a GKE cluster with a GPU node pool:
```sh
gcloud container clusters create kuberay-gpu-cluster \
    --machine-type=g2-standard-24 \
    --location=us-east4-c \
    --num-nodes=2 \
    --accelerator=type=nvidia-l4,count=2,gpu-driver-version=latest
```

This example uses L4 GPUs. Each model replica will use 2 L4 GPUs using vLLM's tensor parallelism.

## Install the KubeRay Operator

> Deploy the KubeRay operator using Helm:

Reviewer 2: Typically, the KubeRay documentation adds a hyperlink to https://docs.ray.io/en/latest/cluster/kubernetes/getting-started/raycluster-quick-start.html#step-2-deploy-a-kuberay-operator when referring to installing a KubeRay operator. Take [RayJob quick start](https://docs.ray.io/en/latest/cluster/kubernetes/getting-started/rayjob-quick-start.html#step-2-install-the-kuberay-operator) as an example,

<img width="759" alt="Screenshot 2024-09-03 at 6 49 29 PM" src="https://github.com/user-attachments/assets/bc9f78a1-f627-44ea-8797-1566e5b8b54a">

Author: Added the link at the bottom of this section. I figured it doesn't hurt to include the steps here so users don't have to open another doc for it

Reviewer 2: The reason I always point to a single location for deploying the KubeRay operator is to make it the sole location for the KubeRay deployment section. This way, when we need to update the documentation, we only need to update one place, avoiding multiple locations that describe how to install the KubeRay operator, which could result in different commands and versions being used.

Author: That makes sense, I'll remove this part and just reference the quick start

Author: updated
